File: src/main/kotlin/nl/boukenijhuis/provider/AbstractOpenAiProvider.java
package nl.boukenijhuis.provider;

import dev.langchain4j.model.chat.ChatModel;
import dev.langchain4j.model.openai.OpenAiChatModel;

public abstract class AbstractOpenAiProvider extends AbstractProvider {

    public AbstractOpenAiProvider(String model) {
        super(model);
    }

    @Override
    public ChatModel getChatModel() {

        return OpenAiChatModel.builder()
                .apiKey(System.getenv(getEnvironmentVariableValue()))
                .baseUrl(getBaseURL())
                .modelName(model)
                // prevents rate limiter logging
                .maxRetries(1)
                .build();
    }

    abstract public String getEnvironmentVariableValue();
    abstract public String getBaseURL();
}

File: src/main/kotlin/nl/boukenijhuis/provider/AbstractOpenAiProvider.kt
package nl.boukenijhuis.provider

import dev.langchain4j.model.chat.ChatModel
import dev.langchain4j.model.openai.OpenAiChatModel

abstract class AbstractOpenAiProvider(model: String) : AbstractProvider(model) {
    override val chatModel: ChatModel
        get() = OpenAiChatModel.builder()
            .apiKey(System.getenv(this.environmentVariableValue))
            .baseUrl(this.baseURL)
            .modelName(model)
            // prevents rate limiter logging
            .maxRetries(0)
            .build()

    abstract val environmentVariableValue: String
    abstract val baseURL: String
}
